node/network/availability-distribution/src/lib.rs

```diff
@@ -0,0 +1,2022 @@
+	view: View,
+) -> Result<()>
+where
+	Context: SubsystemContext<Message = AvailabilityDistributionMessage>,
+{
+	let current = state.peer_views.entry(origin.clone()).or_default();
+
+	let added: Vec<Hash> = view.difference(&*current).cloned().collect();
+
+	*current = view;
+
+	// only contains the intersection of what we are interested and
+	// the union of all relay parent's candidates.
+	let added_candidates = state.cached_live_candidates_unioned(added.iter());
+
+	// Send all messages we've seen before and the peer is now interested
+	// in to that peer.
+
+	for (candidate_hash, _receipt) in added_candidates {
+		let per_candidate = state.per_candidate.entry(candidate_hash).or_default();
+
+		// obtain the relevant chunk indices not sent yet
+		let messages = ((0 as ValidatorIndex)
+			..(per_candidate.validators.len() as ValidatorIndex))
+			.into_iter()
+			.filter_map(|erasure_chunk_index: ValidatorIndex| {
+				let message_id = (candidate_hash, erasure_chunk_index);
+
+				// try to pick up the message from the message vault
+				// so we send as much as we have
+				per_candidate
+					.message_vault
+					.get(&erasure_chunk_index)
+					.filter(|_| {
+						// check if that erasure chunk was already sent before
+						if let Some(sent_set) = per_candidate.sent_messages.get(&origin) {
+							if sent_set.contains(&message_id) {
+								return false;
+							}
+						}
+						true
+					})
+			})
+			.cloned()
+			.collect::<HashSet<_>>();
+
+		send_tracked_gossip_messages_to_peer(ctx, per_candidate, origin.clone(), messages).await?;
+	}
+	Ok(())
+}
+
+/// Obtain the first key with a signing key, which must be ours. We obtain the index as `ValidatorIndex`
+/// If we cannot find a key in the validator set, which we could use.
+fn obtain_our_validator_index(
+	validators: &[ValidatorId],
+	keystore: KeyStorePtr,
+) -> Option<ValidatorIndex> {
+	let keystore = keystore.read();
+	validators.iter().enumerate().find_map(|(idx, validator)| {
+		if keystore.has_keys(&[(validator.to_raw_vec(), PARACHAIN_KEY_TYPE_ID)]) {
+			Some(idx as ValidatorIndex)
+		} else {
+			None
+		}
+	})
+}
+
+/// Handle an incoming message from a peer.
+async fn process_incoming_peer_message<Context>(
+	ctx: &mut Context,
+	state: &mut ProtocolState,
+	origin: PeerId,
+	message: AvailabilityGossipMessage,
+) -> Result<()>
+where
+	Context: SubsystemContext<Message = AvailabilityDistributionMessage>,
+{
+	// obtain the set of candidates we are interested in based on our current view
+	let live_candidates = state.cached_live_candidates_unioned(state.view.0.iter());
+
+	// check if the candidate is of interest
+	let live_candidate = if let Some(live_candidate) = live_candidates.get(&message.candidate_hash)
+	{
+		live_candidate
+	} else {
+		return modify_reputation(ctx, origin, COST_NOT_A_LIVE_CANDIDATE).await;
+	};
+
+	// check the merkle proof
+	let root = &live_candidate.commitments.erasure_root;
+	let anticipated_hash = if let Ok(hash) = branch_hash(
```

Comment:
This can be filed as a TODO for later I guess but we should probably add this checking logic to an `impl` on the `ErasureChunk` primitive.

Reply:
That would probably cause a circular dependency. However, we should expose a better API from `polkadot-erasure-coding`.

Reply:
I'll leave that for another PR